File: backend/controllers/authController.js
import { generateToken } from '../utils/generateToken.js';
import User from '../models/User.js'; 

export const signup = async (req, res) => {
  try {
    const { name, email, password } = req.body;
    if (await User.findOne({ email: email.toLowerCase() })) {
      return res.status(400).json({ msg: 'Email already exists' });
    }
    const user = new User({ name, email: email.toLowerCase(), password });
    await user.save();

    const token = generateToken(user);

    res.cookie('token', token, { 
      httpOnly: true,
      sameSite: 'strict',
      secure: process.env.NODE_ENV === 'production',
      maxAge: 24 * 60 * 60 * 1000
    });

    res.status(201).json({
      msg: 'User created successfully',
      user: {
        id: user._id,
        name: user.name,
        email: user.email,
        avatar: user.avatar
      },
      token
    });
  } catch (error) {
    console.error(error); // This shows up in your terminal!
    res.status(500).json({ msg: error.message || "Server error" });
  }
};

export const login = async (req, res) => {
  try {
    const { email, password } = req.body;
    const user = await User.findOne({ email: email.toLowerCase() });
    if (!user) {
      return res.status(400).json({ msg: 'Invalid email or password' });
    }
    const isMatch = await user.comparePassword(password);
    if (!isMatch) {
      return res.status(400).json({ msg: 'Invalid email or password' });
    }

    const token = generateToken(user);

    res.cookie('token', token, {
      httpOnly: true,
      sameSite: 'strict',
      secure: process.env.NODE_ENV === 'production',
      maxAge: 24 * 60 * 60 * 1000
    });

    res.status(200).json({
      msg: 'Login successful',
      user: {
        id: user._id,
        name: user.name,
        email: user.email,
        avatar: user.avatar
      },
      token
    });
  } catch (error) {
    res.status(500).json({ msg: 'Error logging in', error: error.message });
  }
};

export const logout = (req, res) => {
  res.clearCookie('token', {
    httpOnly: true,
    sameSite: 'strict',
    secure: process.env.NODE_ENV === 'production'
  });
  res.status(200).json({ msg: 'Logout successful' });
};

File: backend-flask/controllers/auth_controller.py
from flask import request, jsonify, make_response
from models.user import User
from utils.token_utils import generate_token
import re
import os


def signup():
    """Handle user signup"""
    try:
        data = request.get_json()
        name = data.get('name')
        email = data.get('email')
        password = data.get('password')

        # Validate email
        if not re.match(r'^[^\s@]+@[^\s@]+\.[^\s@]+$', email):
            return jsonify({'msg': 'Invalid email format'}), 400

        # Check if user exists
        if User.find_by_email(email):
            return jsonify({'msg': 'Email already exists'}), 400

        # Create user
        user = User.create(name, email, password)

        # Generate token
        token = generate_token(user)

        # Create response
        response = make_response(jsonify({
            'msg': 'User created successfully',
            'user': {
                'id': user['id'],
                'name': user['name'],
                'email': user['email'],
                'avatar': user['avatar']
            },
            'token': token
        }), 201)

        # Set cookie
        is_production = os.getenv('NODE_ENV') == 'production'
        response.set_cookie(
            'token',
            token,
            httponly=True,
            samesite='Strict',
            secure=is_production,
            max_age=24 * 60 * 60
        )

        return response

    except Exception as e:
        print(f"Signup error: {str(e)}")
        return jsonify({'msg': str(e) or 'Server error'}), 500


def login():
    """Handle user login"""
    try:
        data = request.get_json()
        email = data.get('email')
        password = data.get('password')

        # Find user
        user = User.find_by_email(email)
        if not user:
            return jsonify({'msg': 'Invalid email or password'}), 400

        # Check password
        if not User.check_password(password, user['password']):
            return jsonify({'msg': 'Invalid email or password'}), 400

        # Generate token
        token = generate_token(user)

        # Create response
        response = make_response(jsonify({
            'msg': 'Login successful',
            'user': {
                'id': user['id'],
                'name': user['name'],
                'email': user['email'],
                'avatar': user['avatar']
            },
            'token': token
        }), 200)

        # Set cookie
        is_production = os.getenv('NODE_ENV') == 'production'
        response.set_cookie(
            'token',
            token,
            httponly=True,
            samesite='Strict',
            secure=is_production,
            max_age=24 * 60 * 60
        )

        return response

    except Exception as e:
        return jsonify({'msg': 'Error logging in', 'error': str(e)}), 500


def logout():
    """Handle user logout"""
    response = make_response(jsonify({'msg': 'Logout successful'}), 200)
    response.set_cookie('token', '', expires=0)
    return response
